src/neuroglancer/datasource/computed/tensorflow/frontend.ts

```diff
@@ -0,0 +1,278 @@
+   * inference loop, if it isn't already running. Returns a promise that
+   * resolves with the prediction output.
+   *
+   * Creates a PendingInference object and inserts it into the correct position
+   * in the inference queue, which is guaranteed to be less than the length of
+   * the backend computation queue in length. As such, it is most likely
+   * faster to simply iterate the queue to insert the request rather than using
+   * a more sophisticated method like binary insertion or a tree-based queue.
+   * @param array the input data to run inference over
+   * @param priority
+   * @param cancellationToken
+   */
+  predict(array: TensorflowArray, priority: number, cancellationToken: CancellationToken):
+      RPCPromise<InferenceResult> {
+    return new Promise((resolve, reject) => {
+             let i;
+             const queue = this.inferenceQueue_;
+             for (i = 0; i < queue.length && queue[i].priority < priority; ++i) {
+             }
+             queue.splice(i, 0, {array, cancellationToken, priority, resolve, reject});
+             this.startInference();
+           })
+        .then((result: InferenceResult) => {
+          return {value: result, transfers: [result.outputBuffer.buffer]};
+        });
+  }
+
+  /**
+   * Start the inference loop, if it isn't already running.
+   */
+  startInference() {
+    if (this.running_) {
+      return;
+    }
+    this.running_ = true;
+    setTimeout(() => this.runInference_(), 0);
+  }
+
+  /**
+   * Run the inference loop recursively.
+   */
+  private runInference_(): Promise<void> {
+    if (this.inferenceQueue_.length === 0) {
+      this.running_ = false;
+      return Promise.resolve();
+    }
+
+    const pendingInference = this.inferenceQueue_.pop()!;
+
+    if (pendingInference.cancellationToken.isCanceled) {
+      pendingInference.reject(CANCELED);
+      return this.runInference_();
+    }
+
+    return this.modelInference_(pendingInference).then(() => this.runInference_());
+  }
+
+  /**
+   * Executes tf.js prediction, resolving or rejecting the pending request's
+   * promise, as appropriate.
+   * @param inferenceRequest  the request to infer.
+   */
+  private modelInference_(inferenceRequest: PendingInference): Promise<void> {
+    const inputLength = inferenceRequest.array.length;
+    const expectedLength = this.params.inputTensorNumElements!;
+    if (inputLength !== expectedLength) {
+      inferenceRequest.reject(
+          new Error(`Input array has ${inputLength} elements. Expected ${expectedLength}`));
+      return Promise.resolve();
+    }
+
+    const prediction = tfjs.tidy(() => {
+      const modelInput = tfjs.tensor(inferenceRequest.array).reshape(this.params.inputTensorShape!);
+      const model = this.model_!;
+      return <Tensor>model.predict(modelInput, {});
+    });
+
+    return prediction.data()
+        .then((outputBuffer: TensorflowArray) => {
+          prediction.dispose();
```

Comment:
I think it would be simpler to do the reference counting on the backend(worker thread) side.

Reply:
(See code review comment on promiseInvoke call)

Reply:
Done. It's also worth mentioning that the .dispose() call here is tfjs-centric. It frees the prediction Tensor's data from the GPU.